File: app/src/main/java/co/edu/unicauca/dopaminah/data/repository/UsageMonitoringRepositoryImpl.kt
package co.edu.unicauca.dopaminah.data.repository

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.*
import androidx.datastore.preferences.preferencesDataStore
import co.edu.unicauca.dopaminah.domain.repository.MonitoringStats
import co.edu.unicauca.dopaminah.domain.repository.UsageMonitoringRepository
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

private val Context.monitoringDataStore: DataStore<Preferences> by preferencesDataStore(name = "usage_monitoring")

@Singleton
class UsageMonitoringRepositoryImpl @Inject constructor(
    @ApplicationContext private val context: Context
) : UsageMonitoringRepository {

    private object PreferencesKeys {
        val TOTAL_SCREEN_TIME = longPreferencesKey("total_screen_time")
        val UNLOCK_COUNT = intPreferencesKey("unlock_count")
        val LAST_RESET_DATE = stringPreferencesKey("last_reset_date")
        val LAST_SCREEN_ON_TIME = longPreferencesKey("last_screen_on_time")
        fun notifiedKey(id: String) = booleanPreferencesKey("notified_$id")
    }

    override fun getMonitoringStats(): Flow<MonitoringStats> {
        return context.monitoringDataStore.data.map { preferences ->
            MonitoringStats(
                totalScreenTimeMillis = preferences[PreferencesKeys.TOTAL_SCREEN_TIME] ?: 0L,
                unlockCount = preferences[PreferencesKeys.UNLOCK_COUNT] ?: 0,
                lastResetDate = preferences[PreferencesKeys.LAST_RESET_DATE] ?: ""
            )
        }
    }

    override suspend fun updateScreenTime(durationMillis: Long) {
        context.monitoringDataStore.edit { preferences ->
            val current = preferences[PreferencesKeys.TOTAL_SCREEN_TIME] ?: 0L
            preferences[PreferencesKeys.TOTAL_SCREEN_TIME] = current + durationMillis
        }
    }

    override suspend fun incrementUnlockCount() {
        context.monitoringDataStore.edit { preferences ->
            val current = preferences[PreferencesKeys.UNLOCK_COUNT] ?: 0
            preferences[PreferencesKeys.UNLOCK_COUNT] = current + 1
        }
    }

    override suspend fun resetDailyStats(date: String) {
        context.monitoringDataStore.edit { preferences ->
            // Clear all notification flags when resetting daily stats
            val keysToRemove = preferences.asMap().keys.filter { it.name.startsWith("notified_") }
            keysToRemove.forEach { preferences.remove(it) }

            preferences[PreferencesKeys.TOTAL_SCREEN_TIME] = 0L
            preferences[PreferencesKeys.UNLOCK_COUNT] = 0
            preferences[PreferencesKeys.LAST_RESET_DATE] = date
        }
    }

    override suspend fun setLastScreenOnTime(timestamp: Long) {
        context.monitoringDataStore.edit { preferences ->
            preferences[PreferencesKeys.LAST_SCREEN_ON_TIME] = timestamp
        }
    }

    override suspend fun getLastScreenOnTime(): Long {
        return context.monitoringDataStore.data.first()[PreferencesKeys.LAST_SCREEN_ON_TIME] ?: 0L
    }

    override suspend fun isAlertNotified(alertId: String): Boolean {
        val today = SimpleDateFormat("yyyyMMdd", Locale.getDefault()).format(Date())
        val key = PreferencesKeys.notifiedKey("${today}_$alertId")
        return context.monitoringDataStore.data.first()[key] ?: false
    }

    override suspend fun markAlertNotified(alertId: String) {
        val today = SimpleDateFormat("yyyyMMdd", Locale.getDefault()).format(Date())
        val key = PreferencesKeys.notifiedKey("${today}_$alertId")
        context.monitoringDataStore.edit { preferences ->
            preferences[key] = true
        }
    }
}

File: DopamiNah/Data/RepositoriesImpl/UsageMonitoringRepositoryImpl.swift
import Foundation

final class UsageMonitoringRepositoryImpl: UsageMonitoringRepositoryProtocol {
    private let defaults: UserDefaults

    init(defaults: UserDefaults = AppGroupHelper.defaults) {
        self.defaults = defaults
    }

    private enum Keys {
        static let totalScreenTime = "total_screen_time"
        static let unlockCount = "unlock_count"
        static let lastResetDate = "last_reset_date"
        static let lastScreenOnTime = "last_screen_on_time"
    }

    func getMonitoringStats() async -> MonitoringStats {
        let totalMillis = Int64(defaults.integer(forKey: Keys.totalScreenTime))
        let unlocks = defaults.integer(forKey: Keys.unlockCount)
        let lastReset = defaults.string(forKey: Keys.lastResetDate) ?? ""
        return MonitoringStats(
            totalScreenTimeMillis: totalMillis,
            unlockCount: unlocks,
            lastResetDate: lastReset
        )
    }

    func updateScreenTime(_ millis: Int64) async {
        let current = defaults.integer(forKey: Keys.totalScreenTime)
        defaults.set(current + Int(millis), forKey: Keys.totalScreenTime)
    }

    func incrementUnlockCount() async {
        let current = defaults.integer(forKey: Keys.unlockCount)
        defaults.set(current + 1, forKey: Keys.unlockCount)
    }

    func resetDailyStats() async {
        defaults.set(0, forKey: Keys.totalScreenTime)
        defaults.set(0, forKey: Keys.unlockCount)
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        defaults.set(formatter.string(from: Date()), forKey: Keys.lastResetDate)

        let allKeys = defaults.dictionaryRepresentation().keys
        for key in allKeys where key.hasPrefix("notified_") {
            defaults.removeObject(forKey: key)
        }
    }

    func setLastScreenOnTime(_ date: Date) async {
        defaults.set(date.timeIntervalSince1970, forKey: Keys.lastScreenOnTime)
    }

    func getLastScreenOnTime() async -> Date? {
        let timestamp = defaults.double(forKey: Keys.lastScreenOnTime)
        guard timestamp > 0 else { return nil }
        return Date(timeIntervalSince1970: timestamp)
    }

    func isAlertNotified(alertId: String) async -> Bool {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let today = formatter.string(from: Date())
        let key = "notified_\(alertId)_\(today)"
        return defaults.bool(forKey: key)
    }

    func markAlertNotified(alertId: String) async {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let today = formatter.string(from: Date())
        let key = "notified_\(alertId)_\(today)"
        defaults.set(true, forKey: key)
    }
}
